// exiftool-vendored/dist/ReadTask.spec.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const geo_tz_1 = __importDefault(require("geo-tz"));
const luxon_1 = require("luxon");
const promises_1 = require("node:fs/promises");
const node_path_1 = __importStar(require("node:path"));
const _chai_spec_1 = require("./_chai.spec");
const DateTime_1 = require("./DateTime");
const ExifDateTime_1 = require("./ExifDateTime");
const ExifTool_1 = require("./ExifTool");
const GeolocationTags_1 = require("./GeolocationTags");
const Object_1 = require("./Object");
const Pick_1 = require("./Pick");
const ReadTask_1 = require("./ReadTask");
const Timezones_1 = require("./Timezones");
const SourceFile = (0, node_path_1.join)((0, _chai_spec_1.tmpdir)(), "example.jpg");
function parse(args) {
    const tt = ReadTask_1.ReadTask.for(args.SourceFile ?? SourceFile, {
        defaultVideosToUTC: true,
        backfillTimezones: true,
        includeImageDataMD5: true,
        ...args,
    });
    const json = JSON.stringify([{ ...args.tags, SourceFile }]);
    // pretend that ExifTool rendered `json`:
    return tt.parse(json, args.error);
}
function geo_tz(lat, lon) {
    try {
        return geo_tz_1.default.find(lat, lon)[0];
    }
    catch {
        return;
    }
}
describe("ReadTask", () => {
    let exiftool;
    before(() => (exiftool = new ExifTool_1.ExifTool()));
    after(() => (0, _chai_spec_1.end)(exiftool));
    describe("Lat/Lon parsing", () => {
        /* Example:
        $ exiftool -j -coordFormat '%.8f' -fast ../test-images/important/Apple_iPhone7Plus.jpg | grep itude
        "GPSLatitudeRef": "North",
        "GPSLongitudeRef": "East",
        "GPSAltitudeRef": "Above Sea Level",
        "GPSAltitude": "73 m Above Sea Level",
        "GPSLatitude": 22.33543889,
        "GPSLongitude": 114.16401667,
       */
        it("N lat is positive", () => {
            (0, _chai_spec_1.expect)(parse({
                tags: {
                    GPSLatitude: 22.33543889,
                    GPSLatitudeRef: "N",
                    GPSLongitude: 1,
                },
            }).GPSLatitude).to.be.closeTo(22.33543889, 0.00001);
        });
        it("S lat is negative", () => {
            (0, _chai_spec_1.expect)(parse({
                tags: {
                    GPSLatitude: -33.84842123,
                    GPSLatitudeRef: "S",
                    GPSLongitude: 1,
                },
            }).GPSLatitude).to.be.closeTo(-33.84842123, 0.00001);
        });
        it("positive E lon is positive", () => {
            (0, _chai_spec_1.expect)(parse({
                tags: {
                    GPSLongitude: 114.16401667,
                    GPSLongitudeRef: "E",
                    GPSLatitude: 1,
                },
            }).GPSLongitude).to.be.closeTo(114.16401667, 0.00001);
        });
        // See https://github.com/photostructure/exiftool-vendored.js/issues/165
        it("negative E lon is negative", () => {
            (0, _chai_spec_1.expect)(parse({
                tags: { GPSLongitude: -114, GPSLongitudeRef: "E", GPSLatitude: 1 },
            })).to.containSubset({
                GPSLongitude: -114,
                GPSLongitudeRef: "W",
            });
        });
        it("positive W lon is negative", () => {
            (0, _chai_spec_1.expect)(parse({
                tags: { GPSLongitude: 122, GPSLongitudeRef: "W", GPSLatitude: 1 },
            })).to.containSubset({
                GPSLongitude: -122,
                GPSLongitudeRef: "W",
            });
        });
        it("negative W lon is positive", () => {
            (0, _chai_spec_1.expect)(parse({
                tags: {
                    GPSLongitude: -122.4406148,
                    GPSLongitudeRef: "W",
                    GPSLatitude: 1,
                },
            }).GPSLongitude).to.be.closeTo(-122.4406148, 0.00001);
        });
        it("parses lat lon even if timezone is given", () => {
            (0, _chai_spec_1.expect)(parse({
                tags: {
                    GPSLongitude: -122.4406148,
                    GPSLongitudeRef: "West",
                    OffsetTime: "+02:00",
                    GPSLatitude: 1,
                },
            }).GPSLongitude).to.be.closeTo(-122.4406148, 0.00001);
        });
        for (const geolocation of [true, false]) {
            for (const preferTimezoneInferenceFromGps of [true, false]) {
                describe(JSON.stringify({ geolocation, preferTimezoneInferenceFromGps }), () => {
                    it("extracts problematic GPSDateTime", async () => {
                        const t = await exiftool.read((0, node_path_1.join)(_chai_spec_1.testDir, "nexus5x.jpg"), {
                            geolocation,
                            preferTimezoneInferenceFromGps,
                        });
                        (0, _chai_spec_1.expect)(t).to.containSubset({
                            MIMEType: "image/jpeg",
                            Make: "LGE",
                            Model: "Nexus 5X",
                            ImageWidth: 16,
                            ImageHeight: 16,
                            tz: "Europe/Zurich",
                            tzSource: geolocation
                                ? "GeolocationTimeZone"
                                : "GPSLatitude/GPSLongitude",
                        });
                        const gpsdt = t.GPSDateTime;
                        (0, _chai_spec_1.expect)(gpsdt.toString()).to.eql("2016-07-19T10:00:24Z");
                        (0, _chai_spec_1.expect)(gpsdt.rawValue).to.eql("2016:07:19 10:00:24Z");
                        (0, _chai_spec_1.expect)(gpsdt.zoneName).to.eql("UTC");
                        if (geolocation) {
                            const actualGeoKeys = Object.keys(t)
                                .filter((ea) => ea.startsWith("Geolocation") &&
                                ea !== "GeolocationWarning")
                                .sort();
                            (0, _chai_spec_1.expect)(actualGeoKeys).to.have.members(GeolocationTags_1.GeolocationTagNames.values.filter((ea) => ea !== "GeolocationWarning"));
                            (0, _chai_spec_1.expect)(t).to.containSubset({
                                GeolocationCity: "Adligenswil",
                                GeolocationRegion: "Lucerne",
                                GeolocationSubregion: "Lucerne-Land District",
                                GeolocationCountryCode: "CH",
                                GeolocationCountry: "Switzerland",
                                GeolocationTimeZone: "Europe/Zurich",
                                GeolocationFeatureCode: "PPL",
                                GeolocationPopulation: 5600,
                                GeolocationPosition: "47.0653 8.3613",
                                GeolocationDistance: "1.71 km",
                                GeolocationBearing: 60,
                            });
                        }
                    });
                });
            }
        }
        it("omits all GPS tags if invalid lat/lon", () => {
            (0, _chai_spec_1.expect)(parse({
                tags: {
                    SourceFile,
                    GPSLatitude: 0,
                    GPSLongitude: 0,
                    GeolocationCity: "Takoradi",
                    GeolocationRegion: "Western",
                    GeolocationSubregion: "Secondi Takoradi",
                    GeolocationCountryCode: "GH",
                    GeolocationCountry: "Ghana",
                    GeolocationTimeZone: "Africa/Accra",
                    GeolocationFeatureCode: "PPL",
                    GeolocationFeatureType: "Populated Place",
                    GeolocationPopulation: 390000,
                    GeolocationPosition: "4.8982, -1.7602",
                    GeolocationDistance: "578.67 km",
                    GeolocationBearing: 340,
                    GPSLatitudeRef: "North",
                    GPSLongitudeRef: "East",
                    GPSPosition: "0 deg 0' 0.00\" N, 0 deg 0' 0.00\" E",
                },
                ignoreZeroZeroLatLon: true,
                geolocation: true,
            })).to.eql({
                SourceFile,
                errors: [],
                warnings: ["Ignoring zero coordinates from GPSLatitude/GPSLongitude"],
            });
        });
        it("extracts GPS tags if valid lat/lon", () => {
            const tags = {
                SourceFile,
                GPSLatitude: "37 deg 48' 3.45\" N",
                GPSLongitude: "122 deg 23' 55.67\" W",
                GPSLatitudeRef: "North",
                GPSLongitudeRef: "West",
                GPSPosition: "37 deg 48' 3.45\" N, 122 deg 23' 55.67\" W",
                GeolocationCity: "Chinatown",
                GeolocationRegion: "California",
                GeolocationSubregion: "City and County of San Francisco",
                GeolocationCountryCode: "US",
                GeolocationCountry: "United States",
                GeolocationTimeZone: "America/Los_Angeles",
                GeolocationFeatureCode: "PPLX",
                GeolocationFeatureType: "Section of Populated Place",
                GeolocationPopulation: 100000,
                GeolocationPosition: "37.7966, -122.4086",
                GeolocationDistance: "1.01 km",
                GeolocationBearing: 246,
            };
            (0, _chai_spec_1.expect)(parse({
                tags,
                ignoreZeroZeroLatLon: true,
                geolocation: true,
            })).to.eql({
                ...tags,
                GPSLatitude: 37.800958,
                GPSLatitudeRef: "N",
                GPSLongitude: -122.398797,
                GPSLongitudeRef: "W",
                zone: "America/Los_Angeles",
                tz: "America/Los_Angeles",
                tzSource: "GeolocationTimeZone",
                errors: [],
                warnings: [],
            });
        });
        describe("without *Ref fields", () => {
            for (const latSign of [1, -1]) {
                for (const lonSign of [1, -1]) {
                    const input = {
                        GPSLatitude: latSign * 34.4,
                        GPSLongitude: lonSign * 119.8,
                    };
                    it(`extracts (${JSON.stringify(input)})`, () => {
                        (0, _chai_spec_1.expect)(parse({ tags: input })).to.containSubset(input);
                    });
                }
            }
        });
    });
    describe("imageHashType", () => {
        const file = (0, node_path_1.join)(_chai_spec_1.testDir, "with_thumb.jpg");
        for (const { imageHashType, hash } of [
            { imageHashType: false, hash: undefined },
            { imageHashType: "MD5", hash: "5617def2642dbd90ab6a2d4f185d7850" },
            {
                imageHashType: "SHA256",
                hash: "5e91b8878501fb77075df22d4056f27e74c4aad8869bcdd2c1cce673b5f23a8d",
            },
            {
                imageHashType: "SHA512",
                hash: "12c4205a228bfba8a77e0da0c4b02dcd381eb15eb06f460781999fd7bc3db2f07f16b0e3a145dfbe68868ee2086e7c47e5b3315bab5146b3f49e7db3d65e1178",
            },
        ]) {
            it(JSON.stringify({ imageHashType }), async () => {
                const t = await exiftool.read(file, undefined, {
                    imageHashType,
                });
                if (hash == null) {
                    (0, _chai_spec_1.expect)(t).to.not.haveOwnProperty("ImageDataHash");
                }
                else {
                    (0, _chai_spec_1.expect)(t).to.haveOwnProperty("ImageDataHash", hash);
                }
            });
        }
    });
    describe("date and time parsing", () => {
        for (const includeMilliseconds of [true, false]) {
            for (const key of ["Time", "DateTimeStamp"]) {
                describe(JSON.stringify({ key, includeMilliseconds }), () => {
                    it("extracts a valid timestamp", () => {
                        const exp = (0, DateTime_1.hms)(luxon_1.DateTime.now(), { includeMilliseconds });
                        const tags = {};
                        tags[key] = exp;
                        const t = parse({ tags });
                        // Seems obvi? well, check out NotDateRe and MaybeDateOrTimeRe.
                        (0, _chai_spec_1.expect)(t[key]).to.be.instanceOf(ExifTool_1.ExifTime);
                        const suffix = key.includes("GPS") ? "+00:00" : "";
                        (0, _chai_spec_1.expect)(t[key]?.toString()).to.eql(exp + suffix);
                    });
                    it("rejects a numeric timestamp", () => {
                        const exp = 12345678;
                        const tags = {};
                        tags[key] = exp;
                        const t = parse({ tags });
                        (0, _chai_spec_1.expect)(t[key]).to.equal(exp);
                    });
                    it("rejects a string timestamp", () => {
                        const exp = "Off";
                        const tags = {};
                        tags[key] = exp;
                        const t = parse({ tags });
                        (0, _chai_spec_1.expect)(t[key]).to.equal(exp);
                    });
                    it("rejects a 00 timestamp", () => {
                        const exp = "00";
                        const tags = {
                            GPSLatitude: 43.7767593,
                            GPSLongitude: 11.2593329,
                        };
                        tags[key] = exp;
                        const t = parse({ tags });
                        (0, _chai_spec_1.expect)(t[key]).to.equal(exp);
                    });
                });
            }
        }
    });
    describe("Time zone extraction", () => {
        it("finds singular positive TimeZoneOffset and sets accordingly", () => {
            const t = parse({
                tags: {
                    TimeZoneOffset: 9,
                    DateTimeOriginal: "2016:08:12 13:28:50",
                },
            });
            (0, _chai_spec_1.expect)(t).to.containSubset({
                tz: "UTC+9",
                tzSource: "TimeZoneOffset",
            });
            (0, _chai_spec_1.expect)(t.DateTimeOriginal).to.containSubset({
                tzoffsetMinutes: 9 * 60,
                zone: "UTC+9",
                inferredZone: true,
            });
        });
        it("respects zero HH:MM OffsetTime (see #203)", () => {
            // this is not UTC, but it is used for "Atlantic/Reykjavik" and other zones
            const t = parse({
                tags: {
                    OffsetTimeOriginal: "+00:00",
                    DateTimeOriginal: "2016:08:12 13:28:50",
                },
            });
            (0, _chai_spec_1.expect)(t.DateTimeOriginal).to.containSubset({
                tzoffsetMinutes: 0,
                inferredZone: true,
            });
            (0, _chai_spec_1.expect)((0, Timezones_1.isUTC)(ExifDateTime_1.ExifDateTime.from(t.DateTimeOriginal)?.zone)).to.eql(true);
        });
        describe("inferTimezoneFromTimeStamp (see #209)", () => {
            it("disabled", () => {
                const t = parse({
                    tags: {
                        DateTimeOriginal: "2016:10:17 09:40:43",
                        CreateDate: "2016:10:17 09:40:43",
                        TimeStamp: "2016:10:17 07:40:43.891-07:00",
                    },
                    inferTimezoneFromTimeStamp: false,
                });
                (0, _chai_spec_1.expect)(t.tz).to.eql(undefined);
            });
            it("enabled", () => {
                const t = parse({
                    tags: {
                        DateTimeOriginal: "2016:10:17 09:40:43",
                        CreateDate: "2016:10:17 09:40:43",
                        TimeStamp: "2016:10:17 07:40:43.891-07:00",
                    },
                    inferTimezoneFromTimeStamp: true,
                });
                (0, _chai_spec_1.expect)(t).to.containSubset({
                    tz: "UTC-5",
                    tzSource: "offset between DateTimeOriginal and TimeStamp",
                });
            });
        });
        it("finds positive array TimeZoneOffset and sets accordingly", () => {
            const t = parse({
                tags: {
                    TimeZoneOffset: [9, 8],
                    DateTimeOriginal: "2016:08:12 13:28:50",
                },
            });
            (0, _chai_spec_1.expect)(t).to.containSubset({
                tz: "UTC+9",
                tzSource: "TimeZoneOffset",
            });
            (0, _chai_spec_1.expect)(t.DateTimeOriginal).to.containSubset({
                tzoffsetMinutes: 9 * 60,
                zone: "UTC+9",
                inferredZone: true,
            });
        });
        it("finds zulu TimeZoneOffset and sets accordingly", () => {
            const t = parse({
                tags: {
                    TimeZoneOffset: 0,
                    DateTimeOriginal: "2016:08:12 13:28:50",
                },
            });
            (0, _chai_spec_1.expect)(t).to.containSubset({
                tz: "UTC",
                tzSource: "TimeZoneOffset",
            });
            (0, _chai_spec_1.expect)(t.DateTimeOriginal).to.containSubset({
                tzoffsetMinutes: 0,
                zone: "UTC",
                inferredZone: true,
            });
        });
        it("finds negative TimeZoneOffset in array and sets accordingly", () => {
            const t = parse({
                tags: {
                    TimeZoneOffset: [-4],
                    DateTimeOriginal: "2016:08:12 13:28:50",
                },
            });
            (0, _chai_spec_1.expect)(t).to.containSubset({
                tz: "UTC-4",
                tzSource: "TimeZoneOffset",
            });
            (0, _chai_spec_1.expect)(t.DateTimeOriginal).to.containSubset({
                tzoffsetMinutes: -4 * 60,
                zone: "UTC-4",
                inferredZone: true,
            });
        });
        it("respects positive HH:MM OffsetTimeOriginal", () => {
            const t = parse({
                tags: {
                    OffsetTimeOriginal: "+03:30",
                    DateTimeOriginal: "2016:08:12 13:28:50",
                },
            });
            (0, _chai_spec_1.expect)(t.DateTimeOriginal).to.containSubset({
                tzoffsetMinutes: 3 * 60 + 30,
                zone: "UTC+3:30",
                inferredZone: true,
            });
        });
        it("respects positive HH OffsetTimeOriginal", () => {
            const t = parse({
                tags: {
                    OffsetTimeOriginal: "+07",
                    DateTimeOriginal: "2016:08:12 13:28:50",
                },
            });
            (0, _chai_spec_1.expect)(t.DateTimeOriginal).to.containSubset({
                tzoffsetMinutes: 7 * 60,
                zone: "UTC+7",
                inferredZone: true,
            });
        });
        it("respects negative HH:MM OffsetTimeOriginal", () => {
            const t = parse({
                tags: {
                    OffsetTimeOriginal: "-09:30",
                    DateTimeOriginal: "2016:08:12 13:28:50",
                },
            });
            (0, _chai_spec_1.expect)(t).to.containSubset({
                tz: "UTC-9:30",
                tzSource: "OffsetTimeOriginal",
            });
            (0, _chai_spec_1.expect)(t.DateTimeOriginal).to.containSubset({
                tzoffsetMinutes: -(9 * 60 + 30),
                year: 2016,
                month: 8,
                day: 12,
                hour: 13,
                minute: 28,
                second: 50,
                millisecond: undefined,
                inferredZone: true,
                zone: "UTC-9:30",
                rawValue: "2016:08:12 13:28:50",
            });
        });
        it("respects negative H OffsetTimeOriginal", () => {
            const t = parse({
                tags: {
                    OffsetTimeOriginal: "-9",
                    DateTimeOriginal: "2016:08:12 13:28:50",
                },
            });
            (0, _chai_spec_1.expect)(t.DateTimeOriginal.tzoffsetMinutes).to.eql(-9 * 60);
            (0, _chai_spec_1.expect)(t.tz).to.eql("UTC-9");
            (0, _chai_spec_1.expect)(t.tzSource).to.eql("OffsetTimeOriginal");
        });
        it("respects negative HH OffsetTimeOriginal", () => {
            const t = parse({
                tags: {
                    OffsetTimeOriginal: "-09",
                    DateTimeOriginal: "2016:08:12 13:28:50",
                },
            });
            (0, _chai_spec_1.expect)(t.DateTimeOriginal.tzoffsetMinutes).to.eql(-9 * 60);
            (0, _chai_spec_1.expect)(t.tz).to.eql("UTC-9");
            (0, _chai_spec_1.expect)(t.tzSource).to.eql("OffsetTimeOriginal");
        });
        it("determines timezone offset from GPS (specifically, Landscape Arch!)", () => {
            const t = parse({
                tags: {
                    GPSLatitude: 38.791121,
                    GPSLatitudeRef: "North",
                    GPSLongitude: -109.606407,
                    GPSLongitudeRef: "West",
                    DateTimeOriginal: "2016:08:12 13:28:50",
                },
            });
            (0, _chai_spec_1.expect)(t.DateTimeOriginal.tzoffsetMinutes).to.eql(-6 * 60);
            (0, _chai_spec_1.expect)(t.tz).to.eql("America/Denver");
            (0, _chai_spec_1.expect)(t.tzSource).to.eql("GPSLatitude/GPSLongitude");
        });
        it("uses GPSDateTime and DateTimeOriginal and sets accordingly for -7", () => {
            const t = parse({
                tags: {
                    DateTimeOriginal: "2016:10:19 11:15:14",
                    GPSDateTime: "2016:10:19 18:15:12",
                    DateTimeCreated: "2016:10:19 11:15:14",
                },
            });
            (0, _chai_spec_1.expect)(t.DateTimeOriginal.tzoffsetMinutes).to.eql(-7 * 60);
            (0, _chai_spec_1.expect)(t.DateTimeCreated.tzoffsetMinutes).to.eql(-7 * 60);
            (0, _chai_spec_1.expect)(t.tz).to.eql("UTC-7");
            (0, _chai_spec_1.expect)(t.tzSource).to.eql("offset between DateTimeOriginal and GPSDateTime");
        });
        it("uses DateTimeUTC and DateTimeOriginal and sets accordingly for +8", () => {
            const t = parse({
                tags: {
                    DateTimeOriginal: "2016:10:19 11:15:14",
                    DateTimeUTC: "2016:10:19 03:15:12",
                    DateTimeCreated: "2016:10:19 11:15:14",
                },
            });
            (0, _chai_spec_1.expect)(t.DateTimeOriginal.tzoffsetMinutes).to.eql(8 * 60);
            (0, _chai_spec_1.expect)(t.DateTimeCreated.tzoffsetMinutes).to.eql(8 * 60);
            (0, _chai_spec_1.expect)(t.tz).to.eql("UTC+8");
            (0, _chai_spec_1.expect)(t.tzSource).to.eql("offset between DateTimeOriginal and DateTimeUTC");
        });
        it("uses DateTimeUTC and DateTimeOriginal and sets accordingly for +5:30", () => {
            const t = parse({
                tags: {
                    DateTimeOriginal: "2018:10:19 11:15:14",
                    DateTimeUTC: "2018:10:19 05:45:12",
                    DateTimeCreated: "2018:10:19 11:15:14",
                },
            });
            (0, _chai_spec_1.expect)(t.DateTimeOriginal.tzoffsetMinutes).to.eql(5.5 * 60);
            (0, _chai_spec_1.expect)(t.DateTimeCreated.tzoffsetMinutes).to.eql(5.5 * 60);
            (0, _chai_spec_1.expect)(t.tz).to.eql("UTC+5:30");
            (0, _chai_spec_1.expect)(t.tzSource).to.eql("offset between DateTimeOriginal and DateTimeUTC");
        });
        it("renders SubSecDateTimeOriginal with no zone if no tz is inferrable", () => {
            const input = {
                DateTimeOriginal: "2016:12:13 09:05:27",
                SubSecDateTimeOriginal: "2016:12:13 09:05:27.12038200",
            };
            const t = parse({ tags: input });
            (0, _chai_spec_1.expect)((0, _chai_spec_1.renderTagsWithRawValues)(t)).to.eql(input);
            (0, _chai_spec_1.expect)((0, _chai_spec_1.renderTagsWithISO)(t)).to.eql({
                DateTimeOriginal: "2016-12-13T09:05:27",
                SubSecDateTimeOriginal: "2016-12-13T09:05:27.120",
                errors: [],
                warnings: [],
            });
        });
        it("renders SubSecDateTimeOriginal for -8", () => {
            const input = {
                DateTimeOriginal: "2016:12:13 09:05:27",
                GPSDateTime: "2016:12:13 17:05:25Z",
                SubSecDateTimeOriginal: "2016:12:13 09:05:27.12038200",
            };
            const t = parse({ tags: input });
            (0, _chai_spec_1.expect)((0, _chai_spec_1.renderTagsWithRawValues)(t)).to.eql(input);
            (0, _chai_spec_1.expect)((0, _chai_spec_1.renderTagsWithISO)(t)).to.eql({
                DateTimeOriginal: "2016-12-13T09:05:27-08:00",
                GPSDateTime: "2016-12-13T17:05:25Z",
                SubSecDateTimeOriginal: "2016-12-13T09:05:27.120-08:00",
                zone: "UTC-8",
                tz: "UTC-8",
                tzSource: "offset between SubSecDateTimeOriginal and GPSDateTime",
                errors: [],
                warnings: [],
            });
        });
        it("skips invalid timestamps", () => {
            const expected = {
                DateTimeOriginal: "2016:08:12 13:28:50",
                GPSDateTime: "not a timestamp",
                SubSecTime: "00",
                SubSecTimeOriginal: "00",
                SubSecTimeDigitized: "00",
            };
            const t = parse({ tags: expected });
            (0, _chai_spec_1.expect)(t).containSubset((0, Object_1.omit)(expected, "DateTimeOriginal"));
            (0, _chai_spec_1.expect)(t.DateTimeOriginal).to.be.instanceOf(ExifDateTime_1.ExifDateTime);
            (0, _chai_spec_1.expect)(t.DateTimeOriginal?.toString()).to.eql("2016-08-12T13:28:50");
            (0, _chai_spec_1.expect)(t.tz).to.eql(undefined);
            (0, _chai_spec_1.expect)(t.tzSource).to.eql(undefined);
        });
        describe("try to reproduce issue #118", () => {
            it("invalid GPSTimeStamp doesn't throw", async () => {
                const t = parse({
                    tags: {
                        GPSTimeStamp: "1970:01:01 00:00:00Z", // < INVALID, this field is always a timestamp without a date
                    },
                });
                (0, _chai_spec_1.expect)(t.GPSTimeStamp.toISOString()).to.eql("1970-01-01T00:00:00Z");
            });
            it("reads file with GPS tags set to common epoch", async () => {
                const t = await exiftool.read((0, node_path_1.join)(_chai_spec_1.testDir, "0epoch.jpg"));
                (0, _chai_spec_1.expect)(t.GPSDateTime.toMillis()).to.eql(0);
            });
        });
        // https://github.com/photostructure/exiftool-vendored.js/issues/113
        describe("timezone parsing", () => {
            const input = {
                TimeZone: "+00:00",
                CreateDate: "2020:08:03 08:00:19-07:00",
                SubSecCreateDate: "2020:08:03 15:00:19.01+00:00",
                DateTimeOriginal: "2020:08:03 15:00:19",
                TimeStamp: "2020:08:03 15:00:19.01",
            };
            it("handles explicit GMT with explicit offset for image/jpeg", () => {
                const MIMEType = "image/jpeg";
                const t = parse({
                    tags: {
                        MIMEType,
                        ...input,
                    },
                });
                (0, _chai_spec_1.expect)((0, _chai_spec_1.renderTagsWithISO)(t)).to.eql({
                    MIMEType,
                    CreateDate: "2020-08-03T08:00:19-07:00",
                    SubSecCreateDate: "2020-08-03T15:00:19.010Z",
                    DateTimeOriginal: "2020-08-03T15:00:19Z",
                    TimeStamp: "2020-08-03T15:00:19.010Z",
                    zone: "UTC",
                    tz: "UTC",
                    tzSource: "TimeZone",
                    TimeZone: "+00:00",
                    errors: [],
                    warnings: [],
                });
            });
            it("handles explicit GMT with explicit offset for video/mp4", () => {
                const MIMEType = "video/mp4/jpeg";
                const t = parse({
                    tags: {
                        MIMEType,
                        ...input,
                    },
                });
                (0, _chai_spec_1.expect)((0, _chai_spec_1.renderTagsWithISO)(t)).to.containSubset({
                    MIMEType,
                    CreateDate: "2020-08-03T08:00:19-07:00", // < because the input had a zone
                    SubSecCreateDate: "2020-08-03T15:00:19.010Z",
                    DateTimeOriginal: "2020-08-03T15:00:19Z",
                    TimeStamp: "2020-08-03T15:00:19.010Z",
                    TimeZone: "+00:00",
                    errors: [],
                    warnings: [],
                });
            });
        });
        describe("iPhone MOV with only CreationDate offset", () => {
            // https://github.com/photostructure/exiftool-vendored.js/issues/151
            it("Timezone from CreationDate with no GPS", () => {
                const t = (0, _chai_spec_1.renderTagsWithISO)(parse({
                    tags: {
                        // exiftool -j -*Zone -MIMEType -*Date -GPS\*\# ~/Downloads/sad.MOV
                        MIMEType: "video/quicktime",
                        FileModifyDate: "2023:07:19 15:38:32-07:00",
                        FileAccessDate: "2023:07:19 15:38:42-07:00",
                        FileInodeChangeDate: "2023:07:19 15:38:42-07:00",
                        CreateDate: "2023:06:11 13:30:35",
                        ModifyDate: "2023:06:11 13:30:46",
                        TrackCreateDate: "2023:06:11 13:30:35",
                        TrackModifyDate: "2023:06:11 13:30:46",
                        MediaCreateDate: "2023:06:11 13:30:35",
                        MediaModifyDate: "2023:06:11 13:30:46",
                        // iPhone 14 Pro renders MOV with only the CreationDate
                        // containing a value offset (!!) -- so the other dates _are
                        // actually in UTC_, not local time (!!!!)
                        CreationDate: "2023:06:11 14:30:35+01:00",
                    },
                    inferTimezoneFromDatestamps: false,
                    backfillTimezones: false,
                }));
                (0, _chai_spec_1.expect)(t).to.eql({
                    zone: "UTC",
                    tz: "UTC",
                    tzSource: "defaultVideosToUTC",
                    MIMEType: "video/quicktime",
                    FileModifyDate: "2023-07-19T15:38:32-07:00",
                    FileAccessDate: "2023-07-19T15:38:42-07:00",
                    FileInodeChangeDate: "2023-07-19T15:38:42-07:00",
                    CreateDate: "2023-06-11T13:30:35Z",
                    ModifyDate: "2023-06-11T13:30:46Z",
                    TrackCreateDate: "2023-06-11T13:30:35Z",
                    TrackModifyDate: "2023-06-11T13:30:46Z",
                    MediaCreateDate: "2023-06-11T13:30:35Z",
                    MediaModifyDate: "2023-06-11T13:30:46Z",
                    CreationDate: "2023-06-11T14:30:35+01:00",
                    errors: [],
                    warnings: [],
                });
            });
            it("Timezone from CreationDate with no GPS and new inferTimezoneFromDatestamps", () => {
                const t = (0, _chai_spec_1.renderTagsWithISO)(parse({
                    tags: {
                        // exiftool -j -*Zone -MIMEType -*Date -GPS\*\# ~/Downloads/sad.MOV
                        MIMEType: "video/quicktime",
                        FileModifyDate: "2023:07:19 15:38:32-07:00",
                        FileAccessDate: "2023:07:19 15:38:42-07:00",
                        FileInodeChangeDate: "2023:07:19 15:38:42-07:00",
                        CreateDate: "2023:06:11 13:30:35",
                        ModifyDate: "2023:06:11 13:30:46",
                        TrackCreateDate: "2023:06:11 13:30:35",
                        TrackModifyDate: "2023:06:11 13:30:46",
                        MediaCreateDate: "2023:06:11 13:30:35",
                        MediaModifyDate: "2023:06:11 13:30:46",
                        // iPhone 14 Pro renders MOV with only the CreationDate
                        // containing a value offset (!!) -- so the other dates _are
                        // actually in UTC_, not local time (!!!!)
                        CreationDate: "2023:06:11 14:30:35+01:00",
                    },
                    inferTimezoneFromDatestamps: true,
                    backfillTimezones: true,
                }));
                (0, _chai_spec_1.expect)(t).to.eql({
                    zone: "UTC+1",
                    tz: "UTC+1",
                    tzSource: "CreationDate",
                    MIMEType: "video/quicktime",
                    FileModifyDate: "2023-07-19T15:38:32-07:00",
                    FileAccessDate: "2023-07-19T15:38:42-07:00",
                    FileInodeChangeDate: "2023-07-19T15:38:42-07:00",
                    CreateDate: "2023-06-11T14:30:35+01:00",
                    ModifyDate: "2023-06-11T14:30:46+01:00",
                    TrackCreateDate: "2023-06-11T14:30:35+01:00",
                    TrackModifyDate: "2023-06-11T14:30:46+01:00",
                    MediaCreateDate: "2023-06-11T14:30:35+01:00",
                    MediaModifyDate: "2023-06-11T14:30:46+01:00",
                    CreationDate: "2023-06-11T14:30:35+01:00",
                    errors: [],
                    warnings: [],
                });
            });
            it("Timezone from CreationDate and GPS", () => {
                const t = (0, _chai_spec_1.renderTagsWithISO)(parse({
                    tags: {
                        // exiftool -j -*Zone -MIMEType -*Date -GPS\*\# ~/Downloads/sad.MOV
                        MIMEType: "video/quicktime",
                        FileModifyDate: "2023:07:19 15:38:32-07:00",
                        FileAccessDate: "2023:07:19 15:38:42-07:00",
                        FileInodeChangeDate: "2023:07:19 15:38:42-07:00",
                        CreateDate: "2023:06:11 13:30:35",
                        ModifyDate: "2023:06:11 13:30:46",
                        TrackCreateDate: "2023:06:11 13:30:35",
                        TrackModifyDate: "2023:06:11 13:30:46",
                        MediaCreateDate: "2023:06:11 13:30:35",
                        MediaModifyDate: "2023:06:11 13:30:46",
                        CreationDate: "2023:06:11 14:30:35+01:00",
                        GPSAltitude: 99.22,
                        GPSAltitudeRef: 0,
                        GPSLatitude: 51.1037,
                        GPSLongitude: -0.8732,
                    },
                    inferTimezoneFromDatestamps: true,
                    backfillTimezones: true,
                }));
                (0, _chai_spec_1.expect)(t).to.eql({
                    MIMEType: "video/quicktime",
                    zone: "Europe/London",
                    tz: "Europe/London",
                    tzSource: "GPSLatitude/GPSLongitude",
                    FileModifyDate: "2023-07-19T15:38:32-07:00",
                    FileAccessDate: "2023-07-19T15:38:42-07:00",
                    FileInodeChangeDate: "2023-07-19T15:38:42-07:00",
                    CreateDate: "2023-06-11T14:30:35+01:00",
                    ModifyDate: "2023-06-11T14:30:46+01:00",
                    TrackCreateDate: "2023-06-11T14:30:35+01:00",
                    TrackModifyDate: "2023-06-11T14:30:46+01:00",
                    MediaCreateDate: "2023-06-11T14:30:35+01:00",
                    MediaModifyDate: "2023-06-11T14:30:46+01:00",
                    CreationDate: "2023-06-11T14:30:35+01:00",
                    GPSAltitude: 99.22,
                    GPSAltitudeRef: 0,
                    GPSLatitude: 51.1037,
                    GPSLongitude: -0.8732,
                    GPSLatitudeRef: "N",
                    GPSLongitudeRef: "W",
                    errors: [],
                    warnings: [],
                });
            });
            it("defaults video without offset to UTC", () => {
                const t = parse({
                    tags: {
                        MIMEType: "video/mp4",
                        CreateDate: "2014:07:17 08:46:27",
                    },
                });
                (0, _chai_spec_1.expect)((0, _chai_spec_1.renderTagsWithISO)(t)).to.eql({
                    // ALL DATES ARE IN ZULU!
                    MIMEType: "video/mp4",
                    CreateDate: "2014-07-17T08:46:27Z",
                    zone: "UTC",
                    tz: "UTC",
                    tzSource: ExifTool_1.defaultVideosToUTC,
                    errors: [],
                    warnings: [],
                });
            });
            it("retains tzoffset in video timestamps", () => {
                const t = parse({
                    tags: {
                        MIMEType: "video/mp4",
                        CreateDate: "2014:07:17 08:46:27-05:00 DST",
                    },
                    // Disable inferTimezoneFromDatestamps to test defaultVideosToUTC behavior
                    inferTimezoneFromDatestamps: false,
                });
                (0, _chai_spec_1.expect)((0, _chai_spec_1.renderTagsWithISO)(t)).to.eql({
                    // ALL DATES ARE IN ZULU!
                    MIMEType: "video/mp4",
                    CreateDate: "2014-07-17T08:46:27-05:00",
                    zone: "UTC",
                    tz: "UTC",
                    tzSource: ExifTool_1.defaultVideosToUTC,
                    errors: [],
                    warnings: [],
                });
            });
            it("handles CET timezone for images", () => {
                const t = parse({
                    tags: {
                        TimeZone: "+01:00",
                        TimeZoneCity: "Rome",
                        CreateDate: "2020:08:03 16:00:19", // < different (local system) zone!
                        SubSecCreateDate: "2020:08:03 16:00:19.123+01:00",
                        DateTimeOriginal: "2020:08:03 16:00:19", // < missing zone!
                        TimeStamp: "2020:08:03 16:00:19.01", // < missing zone!
                    },
                });
                (0, _chai_spec_1.expect)((0, _chai_spec_1.renderTagsWithISO)(t)).to.eql({
                    CreateDate: "2020-08-03T16:00:19+01:00",
                    DateTimeOriginal: "2020-08-03T16:00:19+01:00",
                    SubSecCreateDate: "2020-08-03T16:00:19.123+01:00",
                    TimeStamp: "2020-08-03T16:00:19.010+01:00",
                    TimeZone: "+01:00",
                    TimeZoneCity: "Rome",
                    zone: "UTC+1",
                    tz: "UTC+1",
                    tzSource: "TimeZone",
                    errors: [],
                    warnings: [],
                });
            });
            it("handles CET timezone for video with TimeZone tag", () => {
                const t = (0, _chai_spec_1.renderTagsWithISO)(parse({
                    tags: {
                        MIMEType: "video/mp4",
                        TimeZone: "+01:00",
                        TimeZoneCity: "Rome",
                        CreateDate: "2020:08:03 15:00:19", // < missing zone (assume UTC)
                        DateTimeOriginal: "2020:08:03 15:00:19", // < missing zone (assume UTC)
                        SubSecCreateDate: "2020:08:03 16:00:19.123+01:00",
                        TimeStamp: "2020:08:03 15:00:19.01", // < missing zone (assume UTC)
                    },
                }));
                (0, _chai_spec_1.expect)(t).to.eql({
                    MIMEType: "video/mp4",
                    CreateDate: "2020-08-03T16:00:19+01:00",
                    DateTimeOriginal: "2020-08-03T16:00:19+01:00",
                    SubSecCreateDate: "2020-08-03T16:00:19.123+01:00",
                    TimeStamp: "2020-08-03T16:00:19.010+01:00",
                    TimeZone: "+01:00",
                    TimeZoneCity: "Rome",
                    zone: "UTC+1",
                    tz: "UTC+1",
                    tzSource: "TimeZone",
                    errors: [],
                    warnings: [],
                });
            });
            it("handles CET timezone for video without TimeZone tag", () => {
                const t = parse({
                    tags: {
                        MIMEType: "video/mp4",
                        CreateDate: "2020:08:03 15:00:19", // < naughty video in UTC
                        SubSecCreateDate: "2020:08:03 16:00:19.123+01:00",
                        DateTimeOriginal: "2020:08:03 15:00:19", // < missing zone!
                        TimeStamp: "2020:08:03 15:00:19.01", // < missing zone!
                    },
                    // Disable inferTimezoneFromDatestamps to test defaultVideosToUTC behavior
                    inferTimezoneFromDatestamps: false,
                });
                (0, _chai_spec_1.expect)((0, _chai_spec_1.renderTagsWithISO)(t)).to.eql({
                    MIMEType: "video/mp4",
                    CreateDate: "2020-08-03T15:00:19Z",
                    SubSecCreateDate: "2020-08-03T16:00:19.123+01:00",
                    DateTimeOriginal: "2020-08-03T15:00:19Z",
                    TimeStamp: "2020-08-03T15:00:19.010Z",
                    zone: "UTC",
                    tz: "UTC",
                    tzSource: ExifTool_1.defaultVideosToUTC,
                    errors: [],
                    warnings: [],
                });
            });
            it("doesn't apply missing timezone", () => {
                const t = parse({
                    tags: {
                        // not a video!
                        CreateDate: "2020:08:03 08:00:19-07:00",
                        DateTimeOriginal: "2020:08:03 15:00:19", // < no zone!
                        SubSecCreateDate: "2020:08:03 15:00:19.01+00:00",
                        TimeStamp: "2020:08:03 15:00:19.01", // < no zone!
                    },
                    backfillTimezones: false,
                    // Disable inferTimezoneFromDatestamps to test no-timezone behavior
                    inferTimezoneFromDatestamps: false,
                });
                (0, _chai_spec_1.expect)((0, _chai_spec_1.renderTagsWithISO)(t)).to.eql({
                    // No timezone found, so no normalization:
                    CreateDate: "2020-08-03T08:00:19-07:00",
                    DateTimeOriginal: "2020-08-03T15:00:19", // < no zone!
                    SubSecCreateDate: "2020-08-03T15:00:19.010Z",
                    TimeStamp: "2020-08-03T15:00:19.010", // < no zone!
                    errors: [],
                    warnings: [],
                });
                (0, _chai_spec_1.expect)(t.tz).to.eql(undefined);
                (0, _chai_spec_1.expect)(t.tzSource).to.eql(undefined);
            });
            it("handles EST", () => {
                const t = parse({
                    tags: {
                        CreateDate: "2020:12:29 14:24:45",
                        DateTimeOriginal: "2020:12:29 14:24:45",
                        GPSAltitude: 259.016,
                        GPSDateStamp: "2020:12:29",
                        GPSLatitude: 34.15,
                        GPSLongitude: -84.73,
                        ModifyDate: "2020:12:29 14:24:45",
                        OffsetTime: "-05:00",
                        OffsetTimeDigitized: "-05:00",
                        OffsetTimeOriginal: "-05:00",
                        SubSecCreateDate: "2020:12:29 14:24:45.700-05:00",
                        SubSecDateTimeOriginal: "2020:12:29 14:24:45.700-05:00",
                        SubSecModifyDate: "2020:12:29 14:24:45.789-05:00",
                        SubSecTimeDigitized: 700,
                        SubSecTimeOriginal: 700,
                    },
                });
                (0, _chai_spec_1.expect)((0, _chai_spec_1.renderTagsWithISO)(t)).to.eql({
                    CreateDate: "2020-12-29T14:24:45-05:00",
                    DateTimeOriginal: "2020-12-29T14:24:45-05:00",
                    SubSecCreateDate: "2020-12-29T14:24:45.700-05:00",
                    SubSecDateTimeOriginal: "2020-12-29T14:24:45.700-05:00",
                    SubSecModifyDate: "2020-12-29T14:24:45.789-05:00",
                    ModifyDate: "2020-12-29T14:24:45-05:00",
                    GPSAltitude: 259.016,
                    GPSDateStamp: "2020-12-29",
                    GPSLatitude: 34.15,
                    GPSLongitude: -84.73,
                    GPSLatitudeRef: "N",
                    GPSLongitudeRef: "W",
                    OffsetTime: "-05:00",
                    OffsetTimeDigitized: "-05:00",
                    OffsetTimeOriginal: "-05:00",
                    SubSecTimeDigitized: 700,
                    SubSecTimeOriginal: 700,
                    zone: "UTC-5",
                    tz: "UTC-5",
                    tzSource: "OffsetTimeOriginal",
                    errors: [],
                    warnings: [],
                });
            });
            it("handles EST with only GPS and geo-tz", () => {
                const t = parse({
                    geoTz: geo_tz,
                    tags: {
                        CreateDate: "2020:12:29 14:24:45",
                        GPSLatitude: 34.15,
                        GPSLongitude: -84.73,
                    },
                });
                (0, _chai_spec_1.expect)((0, _chai_spec_1.renderTagsWithISO)(t)).to.eql({
                    CreateDate: "2020-12-29T14:24:45-05:00",
                    GPSLatitude: 34.15,
                    GPSLongitude: -84.73,
                    GPSLatitudeRef: "N",
                    GPSLongitudeRef: "W",
                    zone: "America/New_York",
                    tz: "America/New_York",
                    tzSource: "GPSLatitude/GPSLongitude",
                    errors: [],
                    warnings: [],
                });
            });
            it("{defaultVideosToUTC: true} assumes UTC if video (even if GPS infers EST)", () => {
                const t = parse({
                    tags: {
                        MIMEType: "video/mp4",
                        CreateDate: "2022:08:31 00:32:06",
                        // Smartphone videos seem to always encode timestamps in UTC, even
                        // if there is GPS metadata
                        GPSLatitude: 34.15,
                        GPSLongitude: -84.73,
                    },
                    defaultVideosToUTC: true,
                });
                (0, _chai_spec_1.expect)((0, _chai_spec_1.renderTagsWithISO)(t)).to.eql({
                    MIMEType: "video/mp4",
                    CreateDate: "2022-08-30T20:32:06-04:00",
                    GPSLatitude: 34.15,
                    GPSLongitude: -84.73,
                    GPSLatitudeRef: "N",
                    GPSLongitudeRef: "W",
                    zone: "America/New_York",
                    tz: "America/New_York",
                    tzSource: "GPSLatitude/GPSLongitude",
                    errors: [],
                    warnings: [],
                });
            });
            it("{defaultVideosToUTC: false} assumes video is in local offset (not UTC)", () => {
                const t = parse({
                    tags: {
                        MIMEType: "video/mp4",
                        CreateDate: "2022:08:31 00:32:06", // < FWIW I haven't seen local datestamps in videos in the wild
                        GPSLatitude: 34.15,
                        GPSLongitude: -84.73,
                    },
                    defaultVideosToUTC: false,
                });
                (0, _chai_spec_1.expect)((0, _chai_spec_1.renderTagsWithISO)(t)).to.eql({
                    MIMEType: "video/mp4",
                    CreateDate: "2022-08-31T00:32:06-04:00",
                    GPSLatitude: 34.15,
                    GPSLongitude: -84.73,
                    GPSLatitudeRef: "N",
                    GPSLongitudeRef: "W",
                    zone: "America/New_York",
                    tz: "America/New_York",
                    tzSource: "GPSLatitude/GPSLongitude",
                    errors: [],
                    warnings: [],
                });
            });
            it("infers GPS tz (EDT) if not video", () => {
                const t = parse({
                    tags: {
                        CreateDate: "2022:08:31 00:32:06",
                        GPSLatitude: 34.15,
                        GPSLongitude: -84.73,
                    },
                });
                (0, _chai_spec_1.expect)((0, _chai_spec_1.renderTagsWithISO)(t)).to.eql({
                    CreateDate: "2022-08-31T00:32:06-04:00",
                    GPSLatitude: 34.15,
                    GPSLongitude: -84.73,
                    GPSLatitudeRef: "N",
                    GPSLongitudeRef: "W",
                    zone: "America/New_York",
                    tz: "America/New_York",
                    tzSource: "GPSLatitude/GPSLongitude",
                    errors: [],
                    warnings: [],
                });
            });
            it("normalizes when in EST with only OffsetTimeOriginal", () => {
                const t = parse({
                    tags: {
                        CreateDate: "2020:12:29 14:24:45", // < no zone
                        DateTimeOriginal: "2020:12:29 14:24:45", // < no zone
                        ModifyDate: "2020:12:29 14:24:45", // < no zone
                        OffsetTime: "-05:00",
                        OffsetTimeDigitized: "-05:00",
                        OffsetTimeOriginal: "-05:00",
                        SubSecCreateDate: "2020:12:29 14:24:45.700-05:00",
                        SubSecDateTimeOriginal: "2020:12:29 14:24:45.700-05:00",
                        SubSecModifyDate: "2020:12:29 14:24:45.789-05:00",
                        SubSecTimeDigitized: 700,
                        SubSecTimeOriginal: 700,
                    },
                });
                (0, _chai_spec_1.expect)((0, _chai_spec_1.renderTagsWithISO)(t)).to.eql({
                    // Everything normalized to PST:
                    CreateDate: "2020-12-29T14:24:45-05:00",
                    DateTimeOriginal: "2020-12-29T14:24:45-05:00",
                    ModifyDate: "2020-12-29T14:24:45-05:00",
                    SubSecCreateDate: "2020-12-29T14:24:45.700-05:00",
                    SubSecDateTimeOriginal: "2020-12-29T14:24:45.700-05:00",
                    SubSecModifyDate: "2020-12-29T14:24:45.789-05:00",
                    OffsetTime: "-05:00",
                    OffsetTimeDigitized: "-05:00",
                    OffsetTimeOriginal: "-05:00",
                    SubSecTimeDigitized: 700,
                    SubSecTimeOriginal: 700,
                    zone: "UTC-5",
                    tz: "UTC-5",
                    tzSource: "OffsetTimeOriginal",
                    errors: [],
                    warnings: [],
                });
            });
        });
        describe("ignore UTC zone offsets as valid .tz sources when coming from date and time stamps", () => {
            const input = {
                DateCreated: "2018:07:03",
                TimeCreated: "16:38:32+00:00",
                DateTimeCreated: "2018:07:03 16:38:32+00:00",
                DateTimeOriginal: "2018:07:03 16:38:32",
                FileModifyDate: "2020:01:02 12:34:56-08:00",
            };
            const output = {
                DateCreated: "2018-07-03",
                TimeCreated: "16:38:32+00:00",
                DateTimeCreated: "2018-07-03T16:38:32Z",
                DateTimeOriginal: "2018-07-03T16:38:32",
                FileModifyDate: "2020-01-02T12:34:56-08:00",
                // NOTE! No .tz or .tzSource!
                errors: [],
                warnings: [],
            };
            it("{backfill: false, infer: false} makes no tz changes", () => {
                const t = parse({
                    tags: input,
                    backfillTimezones: false,
                    inferTimezoneFromDatestamps: false,
                });
                (0, _chai_spec_1.expect)((0, _chai_spec_1.renderTagsWithISO)(t)).to.eql(output);
            });
            it("{backfill: true, infer: true|false} adds tz", () => {
                (0, _chai_spec_1.expect)((0, _chai_spec_1.renderTagsWithISO)(parse({
                    tags: input,
                    backfillTimezones: true,
                    inferTimezoneFromDatestamps: false,
                })))
                    .to.eql(output)
                    .and.to.not.haveOwnProperty("tz");
                (0, _chai_spec_1.expect)((0, _chai_spec_1.renderTagsWithISO)(parse({
                    tags: input,
                    backfillTimezones: true,
                    inferTimezoneFromDatestamps: true,
                })))
                    .to.eql(output)
                    .and.to.not.haveOwnProperty("tz");
            });
        });
        describe("issue #156", () => {
            // see https://github.com/photostructure/exiftool-vendored.js/issues/156
            const input = {
                // exiftool -j -*Zone -MIMEType -*Date -GPS\*\# ~/Downloads/File1.mp4
                MIMEType: "video/mp4",
                FileModifyDate: "2022:10:26 16:56:17-07:00",
                FileAccessDate: "2023:09:13 11:44:02-07:00",
                FileInodeChangeDate: "2023:09:13 11:44:02-07:00",
                CreateDate: "2021:09:06 17:44:16",
                ModifyDate: "2021:09:06 17:44:16",
                TrackCreateDate: "2021:09:06 17:44:16",
                TrackModifyDate: "2021:09:06 17:44:16",
                MediaCreateDate: "2021:09:06 17:44:16",
                MediaModifyDate: "2021:09:06 17:44:16",
                GPSCoordinates: "33.3531 -111.8628",
                GPSLatitude: 33.3531,
                GPSLongitude: -111.8628,
                GPSPosition: "33.3531 -111.8628",
            };
            const tz = "America/Phoenix";
            it("{defaultVideosToUTC: true} assumes CreateDate in UTC", () => {
                const t = parse({
                    tags: input,
                    defaultVideosToUTC: true,
                    backfillTimezones: true,
                });
                (0, _chai_spec_1.expect)(t).to.containSubset({
                    MIMEType: "video/mp4",
                    tz,
                    tzSource: "GPSLatitude/GPSLongitude",
                });
                const createDate = t.CreateDate;
                (0, _chai_spec_1.expect)(createDate.zoneName).to.eql(tz);
                // 17:44 from the source is assumed to be in UTC, so we expect 10:44.
                const iso = "2021-09-06T10:44:16-07:00";
                (0, _chai_spec_1.expect)((0, _chai_spec_1.renderTagsWithISO)(t)).to.containSubset({
                    CreateDate: iso,
                    ModifyDate: iso,
                    TrackCreateDate: iso,
                    TrackModifyDate: iso,
                    MediaCreateDate: iso,
                    MediaModifyDate: iso,
                });
            });
            it("{defaultVideosToUTC: false} assumes CreateDate in local time", () => {
                const t = parse({
                    tags: input,
                    defaultVideosToUTC: false,
                    backfillTimezones: true,
                });
                (0, _chai_spec_1.expect)(t).to.containSubset({
                    MIMEType: "video/mp4",
                    tz,
                    tzSource: "GPSLatitude/GPSLongitude",
                });
                const createDate = t.CreateDate;
                (0, _chai_spec_1.expect)(createDate.zoneName).to.eql(tz);
                // note the 17:00 hour from the source is not changed by inheriting
                // the TZ from the GPS location:
                const iso = "2021-09-06T17:44:16-07:00";
                (0, _chai_spec_1.expect)((0, _chai_spec_1.renderTagsWithISO)(t)).to.containSubset({
                    CreateDate: iso,
                    ModifyDate: iso,
                    TrackCreateDate: iso,
                    TrackModifyDate: iso,
                    MediaCreateDate: iso,
                    MediaModifyDate: iso,
                });
            });
            it("{defaultVideosToUTC: false, backfillTimezones: false}", () => {
                const t = parse({
                    tags: input,
                    defaultVideosToUTC: false,
                    backfillTimezones: false,
                });
                (0, _chai_spec_1.expect)(t).to.containSubset({
                    ...(0, Pick_1.pick)(input, "MIMEType", "GPSCoordinates", "GPSLatitude", "GPSLongitude"),
                    tz,
                    tzSource: "GPSLatitude/GPSLongitude",
                });
                const createDate = t.CreateDate;
                (0, _chai_spec_1.expect)(createDate.zoneName).to.eql(undefined);
                (0, _chai_spec_1.expect)(createDate.inferredZone).to.eql(false);
                // note the 17:00 hour from the source is not changed by inheriting
                // the TZ from the GPS location:
                const iso = "2021-09-06T17:44:16";
                (0, _chai_spec_1.expect)((0, _chai_spec_1.renderTagsWithISO)(t)).to.containSubset({
                    CreateDate: iso,
                    ModifyDate: iso,
                    TrackCreateDate: iso,
                    TrackModifyDate: iso,
                    MediaCreateDate: iso,
                    MediaModifyDate: iso,
                });
            });
        });
    });
    it("Resource.When is parsed by ExifDateTime", () => {
        const t = parse({
            tags: {
                History: [
                    {
                        Action: "infer",
                        Changed: "Make",
                        Parameters: '"ppeggj"',
                        SoftwareAgent: "PhotoStructure",
                        When: "2023:10:01 17:13:07.141",
                    },
                ],
            },
        });
        (0, _chai_spec_1.expect)(t.History).to.be.an("array");
        const w = t.History[0].When;
        (0, _chai_spec_1.expect)(w).to.be.instanceof(ExifDateTime_1.ExifDateTime);
        (0, _chai_spec_1.expect)(w.toString()).to.eql("2023-10-01T17:13:07.141");
    });
    describe("SubSecDateTimeOriginal", () => {
        it("extracts datetimestamp with millis", () => {
            const t = parse({
                tags: { SubSecDateTimeOriginal: "2016:10:19 11:15:14.437831" },
            }).SubSecDateTimeOriginal;
            (0, _chai_spec_1.expect)(t.year).to.eql(2016);
            (0, _chai_spec_1.expect)(t.month).to.eql(10);
            (0, _chai_spec_1.expect)(t.day).to.eql(19);
            (0, _chai_spec_1.expect)(t.hour).to.eql(11);
            (0, _chai_spec_1.expect)(t.minute).to.eql(15);
            (0, _chai_spec_1.expect)(t.second).to.eql(14);
            (0, _chai_spec_1.expect)(t.tzoffsetMinutes).to.eql(undefined);
            (0, _chai_spec_1.expect)(t.millisecond).to.eql(437);
            const d = t.toDate();
            (0, _chai_spec_1.expect)(d.getFullYear()).to.eql(2016);
            (0, _chai_spec_1.expect)(d.getMonth()).to.eql(10 - 1);
            (0, _chai_spec_1.expect)(d.getDate()).to.eql(19);
            (0, _chai_spec_1.expect)(d.getHours()).to.eql(11);
            (0, _chai_spec_1.expect)(d.getMinutes()).to.eql(15);
            (0, _chai_spec_1.expect)(d.getSeconds()).to.eql(14);
            (0, _chai_spec_1.expect)(d.getMilliseconds()).to.eql(437); // Javascript Date doesn't do fractional millis.
        });
    });
    describe("EXIFTOOL_HOME", () => {
        let et;
        before(() => (et = new ExifTool_1.ExifTool({
            exiftoolEnv: {
                EXIFTOOL_HOME: node_path_1.default.resolve(__dirname, "..", "test"),
            },
        })));
        after(() => (0, _chai_spec_1.end)(et));
        it("returns the new custom tag", async () => {
            const t = await et.read("./test/pixel.jpg");
            // This is a non-standard tag, added by the custom user configuration:
            (0, _chai_spec_1.expect)(t.UppercaseBaseName).to.eql("PIXEL");
        });
    });
    describe("non-alphanumeric filenames", () => {
        for (const { str, desc } of _chai_spec_1.NonAlphaStrings) {
            it("reads with " + desc, async () => {
                const FileName = str + ".jpg";
                const dest = node_path_1.default.join((0, _chai_spec_1.tmpdir)(), FileName);
                await (0, _chai_spec_1.mkdirp)((0, _chai_spec_1.tmpdir)());
                await (0, promises_1.copyFile)(node_path_1.default.join(_chai_spec_1.testDir, "quotes.jpg"), dest);
                const t = await exiftool.read(dest);
                (0, _chai_spec_1.expect)(t).to.containSubset({
                    MIMEType: "image/jpeg",
                    FileName,
                    Make: "Apple",
                    Model: "iPhone 7 Plus",
                    ImageDescription: "image description for quotes test",
                    LastKeywordXMP: ["Test", "examples", "beach"],
                    Title: _chai_spec_1.UnicodeTestMessage,
                });
                (0, _chai_spec_1.expect)(t.DateTimeOriginal?.toString()).to.match(/^2016-08-12T13:28:50/);
            });
        }
    });
    /**
     * @see https://github.com/photostructure/exiftool-vendored.js/issues/157
     */
    describe("issue 157", () => {
        it("time stamp parsing without inferTimezoneFromDatestamps", () => {
            const t = parse({
                inferTimezoneFromDatestamps: false,
                backfillTimezones: false,
                tags: {
                    DateTimeOriginal: "2014:04:06 00:47:40",
                    TimeCreated: "00:47:40+02:00",
                },
            });
            (0, _chai_spec_1.expect)(t.tz).to.eql(undefined);
            (0, _chai_spec_1.expect)(t.DateTimeOriginal).to.be.instanceof(ExifDateTime_1.ExifDateTime);
            (0, _chai_spec_1.expect)(t.DateTimeOriginal?.toString()).to.eql("2014-04-06T00:47:40");
            (0, _chai_spec_1.expect)(t.DateTimeOriginal).to.containSubset({
                year: 2014,
                month: 4,
                day: 6,
                hour: 0,
                minute: 47,
                second: 40,
                millisecond: undefined,
                inferredZone: false,
                zone: undefined,
                rawValue: "2014:04:06 00:47:40",
            });
            (0, _chai_spec_1.expect)(t.TimeCreated).to.be.instanceof(ExifTool_1.ExifTime);
            (0, _chai_spec_1.expect)(t.TimeCreated?.toString()).to.eql("00:47:40+02:00");
            (0, _chai_spec_1.expect)(t.TimeCreated).to.containSubset({
                hour: 0,
                minute: 47,
                second: 40,
                millisecond: undefined,
                inferredZone: false,
                zone: "UTC+2",
                rawValue: "00:47:40+02:00",
            });
        });
        it("timezone offset is pulled from timestamps with offsets", () => {
            const t = parse({
                inferTimezoneFromDatestamps: true,
                backfillTimezones: false,
                tags: {
                    DateTimeOriginal: "2014:04:06 00:47:40",
                    TimeCreated: "00:47:40+02:00",
                },
            });
            (0, _chai_spec_1.expect)(t).to.containSubset({
                tz: "UTC+2",
                tzSource: "TimeCreated",
            });
            (0, _chai_spec_1.expect)(t.DateTimeOriginal).to.be.instanceof(ExifDateTime_1.ExifDateTime);
            (0, _chai_spec_1.expect)(t.DateTimeOriginal?.toString()).to.eql("2014-04-06T00:47:40");
            (0, _chai_spec_1.expect)(t.DateTimeOriginal).to.containSubset({
                year: 2014,
                month: 4,
                day: 6,
                hour: 0,
                minute: 47,
                second: 40,
                millisecond: undefined,
                inferredZone: false,
                zone: undefined,
                rawValue: "2014:04:06 00:47:40",
            });
            (0, _chai_spec_1.expect)(t.TimeCreated).to.be.instanceof(ExifTool_1.ExifTime);
            (0, _chai_spec_1.expect)(t.TimeCreated?.toString()).to.eql("00:47:40+02:00");
            (0, _chai_spec_1.expect)(t.TimeCreated).to.containSubset({
                hour: 0,
                minute: 47,
                second: 40,
                millisecond: undefined,
                inferredZone: false,
                zone: "UTC+2",
                rawValue: "00:47:40+02:00",
            });
        });
        it("timezone offset is pulled from timestamps with offsets and backfills", () => {
            const t = parse({
                inferTimezoneFromDatestamps: true,
                backfillTimezones: true,
                tags: {
                    DateTimeOriginal: "2014:04:06 00:47:40",
                    TimeCreated: "00:47:40+02:00",
                },
            });
            (0, _chai_spec_1.expect)(t).to.containSubset({
                tz: "UTC+2",
                tzSource: "TimeCreated",
            });
            (0, _chai_spec_1.expect)(t.DateTimeOriginal).to.be.instanceof(ExifDateTime_1.ExifDateTime);
            (0, _chai_spec_1.expect)(t.DateTimeOriginal?.toString()).to.eql("2014-04-06T00:47:40+02:00");
            (0, _chai_spec_1.expect)(t.DateTimeOriginal).to.containSubset({
                year: 2014,
                month: 4,
                day: 6,
                hour: 0,
                minute: 47,
                second: 40,
                millisecond: undefined,
                inferredZone: true,
                zone: "UTC+2",
                rawValue: "2014:04:06 00:47:40",
            });
            (0, _chai_spec_1.expect)(t.TimeCreated).to.be.instanceof(ExifTool_1.ExifTime);
            (0, _chai_spec_1.expect)(t.TimeCreated?.toString()).to.eql("00:47:40+02:00");
            (0, _chai_spec_1.expect)(t.TimeCreated).to.containSubset({
                hour: 0,
                minute: 47,
                second: 40,
                millisecond: undefined,
                inferredZone: false,
                zone: "UTC+2",
                rawValue: "00:47:40+02:00",
            });
        });
    });
    /**
     * @see https://github.com/photostructure/exiftool-vendored.js/issues/215
     */
    describe("issue #215", () => {
        for (const geolocation of [true, false]) {
            for (const preferTimezoneInferenceFromGps of [true, false]) {
                describe(JSON.stringify({ geolocation, preferTimezoneInferenceFromGps }), () => {
                    it("adjusts Nikon timezones by an hour if DaylightSavings is truthy", () => {
                        // From https://github.com/immich-app/immich/issues/13141#issuecomment-2390788790
                        // exiftool -j -struct -GPS*# ~/src/exiftool-vendored.js/test/nikon-daylight-savings.jpg -\*time\* -\*date\* -Daylight* -\*zone\* -Make -Model
                        const t = parse({
                            geolocation,
                            preferTimezoneInferenceFromGps,
                            backfillTimezones: true,
                            tags: {
                                GPSLatitude: -45.8745231666667,
                                GPSLongitude: 170.503112783333,
                                GPSLatitudeRef: "S",
                                GPSLongitudeRef: "E",
                                GPSPosition: "-45.8745231666667 170.503112783333",
                                ExposureTime: "1/20",
                                OffsetTime: "+13:00",
                                OffsetTimeOriginal: "+13:00",
                                OffsetTimeDigitized: "+13:00",
                                TimeZone: "+12:00",
                                ISOAutoShutterTime: "Auto",
                                SelfTimerTime: "10 s",
                                SelfTimerShotCount: 9,
                                SelfTimerShotInterval: "0.5 s",
                                PlaybackMonitorOffTime: "10 s",
                                MenuMonitorOffTime: "1 min",
                                ShootingInfoMonitorOffTime: "4 s",
                                ImageReviewMonitorOffTime: "4 s",
                                LiveViewMonitorOffTime: "10 min",
                                PowerUpTime: "0000:00:00 00:00:00",
                                SubSecTime: 45,
                                SubSecTimeDigitized: 45,
                                DateTimeOriginal: "2020:02:10 20:24:43",
                                ProfileDateTime: "2024:10:01 13:21:03",
                                SubSecDateTimeOriginal: "2020:02:10 20:24:43+13:00",
                                FileModifyDate: "2024:10:05 11:34:57-07:00",
                                FileAccessDate: "2024:10:05 11:34:57-07:00",
                                FileInodeChangeDate: "2024:10:05 11:34:57-07:00",
                                ModifyDate: "2024:10:03 08:15:25",
                                CreateDate: "2020:02:10 20:24:43",
                                DateDisplayFormat: "D/M/Y",
                                SubSecCreateDate: "2020:02:10 20:24:43.45", // < needs backfill
                                SubSecModifyDate: "2024:10:03 08:15:25.45+13:00",
                                DaylightSavings: "Yes",
                                Make: "NIKON CORPORATION",
                                Model: "NIKON D7500",
                            },
                        });
                        (0, _chai_spec_1.expect)((0, _chai_spec_1.renderTagsWithISO)(t)).to.containSubset({
                            DateTimeOriginal: "2020-02-10T20:24:43+13:00",
                            SubSecDateTimeOriginal: "2020-02-10T20:24:43+13:00",
                            SubSecCreateDate: "2020-02-10T20:24:43.450+13:00",
                            SubSecModifyDate: "2024-10-03T08:15:25.450+13:00",
                            ...(preferTimezoneInferenceFromGps
                                ? {
                                    tz: "Pacific/Auckland",
                                    tzSource: "GPSLatitude/GPSLongitude",
                                }
                                : {
                                    tz: "UTC+13",
                                    tzSource: "TimeZone (adjusted for DaylightSavings)",
                                }),
                        });
                    });
                });
            }
        }
    });
    describe("GPS coordinate parsing", () => {
        it("should parse valid GPS coordinates", () => {
            const src = (0, _chai_spec_1.tmpname)();
            const t = ReadTask_1.ReadTask.for(src, {});
            const result = t.parse(JSON.stringify([
                {
                    SourceFile: src,
                    GPSPosition: "37 deg 46' 29.64\" N, 122 deg 25' 9.84\" W",
                },
            ]));
            (0, _chai_spec_1.expect)(result).to.containSubset({
                GPSLatitude: 37.7749,
                GPSLongitude: -122.4194,
            });
        });
    });
    describe("Apple Photos exports", () => {
        it("parses XMP #1", async () => {
            const t = await exiftool.read("test/iphone-export-1.xmp", {
                geolocation: true,
            });
            (0, _chai_spec_1.expect)(t).to.containSubset({
                GPSLatitude: 44.646258,
                GPSLongitude: -63.569717,
                GPSLatitudeRef: "N",
                GPSLongitudeRef: "W",
                GeolocationPosition: "44.6483 -63.5775",
                GPSPosition: "44.6462583333333 63.5697166666667",
                tz: "America/Halifax",
                tzSource: "GeolocationTimeZone",
            });
        });
        it("parses XMP #2", async () => {
            const t = await exiftool.read("test/iphone-export-2.xmp", {
                geolocation: true,
            });
            (0, _chai_spec_1.expect)(t).to.containSubset({
                Title: "Title Photos 7 åäöÅÄÖ",
                Description: "Caption Photos 7 åäöÅÄÖ",
                Subject: ["Keyword 1 Photos 7 åäöÅÄÖ", "Keyword 2 Photos 7 åäöÅÄÖ"],
                GPSLatitude: -33.451856,
                GPSLatitudeRef: "S",
                GPSLongitude: -70.650466,
                GPSLongitudeRef: "W",
                GeolocationPosition: "-33.4570 -70.6483",
                GPSPosition: "33.4518558 70.6504661",
                tz: "America/Santiago",
                tzSource: "GeolocationTimeZone",
            });
        });
    });
    describe("ExifToolVersion parsing", () => {
        // We can't use our parse() function here, because we need specific control
        // of the JSON formatting (ExifTool outputs significant trailing zeros in
        // the version string!)
        it("preserves version string with single decimal place", () => {
            const src = (0, _chai_spec_1.tmpname)();
            const tt = ReadTask_1.ReadTask.for(src, {});
            const json = `[{"SourceFile":${JSON.stringify(src)},"ExifToolVersion":12.3}]`;
            const result = tt.parse(json);
            (0, _chai_spec_1.expect)(result.ExifToolVersion).to.be.a("string");
            (0, _chai_spec_1.expect)(result.ExifToolVersion).to.eql("12.3");
        });
        it("preserves version string with two decimal places", () => {
            const src = (0, _chai_spec_1.tmpname)();
            const tt = ReadTask_1.ReadTask.for(src, {});
            const json = `[{"SourceFile":${JSON.stringify(src)},"ExifToolVersion":12.30}]`;
            const result = tt.parse(json);
            (0, _chai_spec_1.expect)(result.ExifToolVersion).to.be.a("string");
            (0, _chai_spec_1.expect)(result.ExifToolVersion).to.eql("12.30");
        });
        it("handles version with no decimal places", () => {
            const src = (0, _chai_spec_1.tmpname)();
            const tt = ReadTask_1.ReadTask.for(src, {});
            const json = `[{"SourceFile":${JSON.stringify(src)},"ExifToolVersion":12}]`;
            const result = tt.parse(json);
            (0, _chai_spec_1.expect)(result.ExifToolVersion).to.be.a("string");
            (0, _chai_spec_1.expect)(result.ExifToolVersion).to.eql("12");
        });
        it("matches version from exiftool.version() method", async () => {
            const tags = await exiftool.read((0, node_path_1.join)(_chai_spec_1.testDir, "img.jpg"));
            const version = await exiftool.version();
            // Both should be strings and match
            (0, _chai_spec_1.expect)(tags.ExifToolVersion).to.be.a("string");
            (0, _chai_spec_1.expect)(version).to.be.a("string");
            (0, _chai_spec_1.expect)(tags.ExifToolVersion).to.eql(version);
        });
        it("handles different JSON whitespace formatting", () => {
            const src = (0, _chai_spec_1.tmpname)();
            const tt = ReadTask_1.ReadTask.for(src, {});
            // Test different whitespace scenarios
            const testCases = [
                '{"ExifToolVersion":12.3}',
                '{"ExifToolVersion": 12.3}',
                '{"ExifToolVersion" : 12.3}',
                '{"ExifToolVersion"    :    12.3}',
            ];
            testCases.forEach((jsonStr) => {
                const fullJson = `[${jsonStr.replace("}", ',"SourceFile":' + JSON.stringify(src) + "}")}]`;
                const result = tt.parse(fullJson);
                (0, _chai_spec_1.expect)(result.ExifToolVersion).to.be.a("string");
                (0, _chai_spec_1.expect)(result.ExifToolVersion).to.eql("12.3");
            });
        });
    });
    describe("MWG Face Region parsing", () => {
        const faceFile = (0, node_path_1.join)(_chai_spec_1.testDir, "with_faces.jpg");
        it("reads RegionInfo as nested struct with default struct=1", async () => {
            const t = await exiftool.read(faceFile);
            (0, _chai_spec_1.expect)(t.RegionInfo).to.be.an("object");
            const regionInfo = t.RegionInfo;
            (0, _chai_spec_1.expect)(regionInfo.AppliedToDimensions).to.containSubset({
                W: 640,
                H: 480,
                Unit: "pixel",
            });
            (0, _chai_spec_1.expect)(regionInfo.RegionList).to.be.an("array").with.lengthOf(2);
            (0, _chai_spec_1.expect)(regionInfo.RegionList[0]).to.containSubset({
                Name: "Alice Smith",
                Type: "Face",
            });
            (0, _chai_spec_1.expect)(regionInfo.RegionList[0].Area).to.containSubset({
                X: 0.35,
                Y: 0.4,
                W: 0.25,
                H: 0.35,
                Unit: "normalized",
            });
            (0, _chai_spec_1.expect)(regionInfo.RegionList[1]).to.containSubset({
                Name: "Bob Jones",
                Type: "Face",
            });
        });
        it("reads flattened region tags with struct=0", async () => {
            const t = await exiftool.read(faceFile, { struct: 0 });
            // With struct=0, RegionInfo becomes flattened tags
            (0, _chai_spec_1.expect)(t.RegionInfo).to.eql(undefined);
            // Check flattened tags exist
            const tags = t;
            (0, _chai_spec_1.expect)(tags.RegionAppliedToDimensionsW).to.eql(640);
            (0, _chai_spec_1.expect)(tags.RegionAppliedToDimensionsH).to.eql(480);
            (0, _chai_spec_1.expect)(tags.RegionAppliedToDimensionsUnit).to.eql("pixel");
            // Multiple faces = arrays
            (0, _chai_spec_1.expect)(tags.RegionName).to.eql(["Alice Smith", "Bob Jones"]);
            (0, _chai_spec_1.expect)(tags.RegionType).to.eql(["Face", "Face"]);
            (0, _chai_spec_1.expect)(tags.RegionAreaX).to.eql([0.35, 0.7]);
            (0, _chai_spec_1.expect)(tags.RegionAreaY).to.eql([0.4, 0.45]);
        });
        it("reads PersonInImage as array", async () => {
            const t = await exiftool.read(faceFile);
            (0, _chai_spec_1.expect)(t.PersonInImage).to.be.an("array");
            (0, _chai_spec_1.expect)(t.PersonInImage).to.eql(["Alice Smith", "Bob Jones"]);
        });
    });
});
