// batch-cluster/dist/exit-backstop-helper.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Subprocess helper for the exit-backstop TDD test in BatchCluster.spec.ts.
 *
 * Flow:
 *  1. Spawn a BatchCluster with one child process (test.js with IGNORE_EXIT=1).
 *  2. Run `upcase hello` to confirm the process is healthy.
 *  3. Run `keepalive 60000` — test.js starts a 60-second ref'd timer then
 *     immediately responds with PASS.  By awaiting this task we know the timer
 *     is running, so the child process will survive even after stdin closes
 *     (which happens when the helper exits and the OS closes the pipe).
 *  4. Print the live PID to stdout for the parent test to collect.
 *  5. Call bc.end(true) fire-and-forget, then process.exit(0) immediately —
 *     before the async cleanup chain can run.
 *
 * The child process stays alive (kept by the 60-second timer) unless something
 * explicitly SIGKILLs it.  That "something" is the exit backstop in BatchCluster.
 *
 *   Without fix → PID survives (orphaned): parent test FAILS
 *   With fix    → SIGKILL from backstop kills PID: parent test PASSES
 */
const node_child_process_1 = __importDefault(require("node:child_process"));
const node_path_1 = __importDefault(require("node:path"));
const node_process_1 = __importDefault(require("node:process"));
const BatchCluster_1 = require("./BatchCluster");
const Parser_1 = require("./Parser");
const Task_1 = require("./Task");
// IGNORE_EXIT: "1" so test.js ignores the "exit" command and SIGTERM.
// Only SIGKILL (from the backstop fix) can terminate it.
const env = {
    FAIL_RATE: "0",
    RNG_SEED: "backstop-helper",
    NEWLINE: "lf",
    IGNORE_EXIT: "1",
    UNLUCKY_FAIL: "0",
};
const bc = new BatchCluster_1.BatchCluster({
    processFactory: () => node_child_process_1.default.spawn(node_process_1.default.execPath, [node_path_1.default.join(__dirname, "test.js")], {
        env,
    }),
    versionCommand: "version",
    pass: "PASS",
    fail: "FAIL",
    exitCommand: "exit",
    maxProcs: 1,
    cleanupChildProcsOnExit: true,
    spawnTimeoutMillis: 10_000,
    taskTimeoutMillis: 10_000,
});
async function main() {
    // Confirm the process is healthy.
    await bc.enqueueTask(new Task_1.Task("upcase hello", Parser_1.SimpleParser));
    // "keepalive 60000" starts a 60-second ref'd timer in test.js then responds
    // immediately with PASS.  Awaiting here guarantees the timer is running
    // before we continue — so the process outlives the stdin-EOF that happens
    // when this helper process exits.
    await bc.enqueueTask(new Task_1.Task("keepalive 60000", Parser_1.SimpleParser));
    const pids = bc.pids();
    if (pids.length === 0) {
        node_process_1.default.stderr.write("ERROR: no live PIDs after keepalive task\n");
        node_process_1.default.exit(1);
    }
    // Write live PIDs so the parent test can read them.
    for (const pid of pids)
        node_process_1.default.stdout.write(pid + "\n");
    node_process_1.default.stdout.write("PIDS_DONE\n");
    // Fire-and-forget: async cleanup won't run before process.exit(0).
    void bc.end(true);
    node_process_1.default.exit(0);
}
main().catch((err) => {
    node_process_1.default.stderr.write(String(err) + "\n");
    node_process_1.default.exit(1);
});
