// batch-cluster/dist/find-flush-thresholds.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const FindFlushThresholds_1 = require("./FindFlushThresholds");
const FlushThresholdTestHelpers_1 = require("./FlushThresholdTestHelpers");
const Platform_1 = require("./Platform");
const Task_1 = require("./Task");
// ---------------------------------------------------------------------------
// CLI (internal testing tool — only runs when executed directly)
// ---------------------------------------------------------------------------
/* eslint-disable no-console */
const platformName = Platform_1.isWin ? "windows" : Platform_1.isMac ? "macos" : "linux";
const cliLogger = {
    trace: () => { }, // eslint-disable-line @typescript-eslint/no-empty-function
    debug: () => { }, // eslint-disable-line @typescript-eslint/no-empty-function
    info: console.log,
    warn: console.warn,
    error: console.error,
};
function printUsage() {
    console.log(`Usage: node find-flush-thresholds.js [options]

Options:
  --lo <ms>            Lower bound (default: 0)
  --hi <ms>            Upper bound (default: 500)
  --help               Show this help`);
}
function parseArgs() {
    const args = process.argv.slice(2);
    const config = { lo: 0, hi: 500 };
    for (let i = 0; i < args.length; i++) {
        const arg = args[i];
        const next = () => {
            const v = args[++i];
            if (v == null) {
                console.error(`Missing value for ${arg}`);
                process.exit(1);
            }
            return v;
        };
        switch (arg) {
            case "--lo":
                config.lo = parseInt(next(), 10);
                break;
            case "--hi":
                config.hi = parseInt(next(), 10);
                break;
            case "--help":
                printUsage();
                process.exit(0);
            // eslint-disable-next-line no-fallthrough
            default:
                console.error(`Unknown argument: ${arg}`);
                printUsage();
                process.exit(1);
        }
    }
    return config;
}
async function main() {
    const config = parseArgs();
    console.log(`Finding streamFlushMillis threshold on ${platformName} (node ${process.version})`);
    const recommended = await (0, FindFlushThresholds_1.findStreamFlushMillis)({
        processFactory: FlushThresholdTestHelpers_1.testProcessFactory,
        versionCommand: "version",
        pass: "PASS",
        fail: "FAIL",
        exitCommand: "exit",
        taskFactory: (i) => new Task_1.Task("stderrfail test-data " + i, FlushThresholdTestHelpers_1.expectFailParser),
        lo: config.lo,
        hi: config.hi,
        logger: () => cliLogger,
    });
    console.log("\n" + "=".repeat(60));
    console.log(`streamFlushMillis on ${platformName} (${process.version}):`);
    console.log(`  Minimum reliable: ${recommended}ms`);
    console.log("=".repeat(60));
}
main().catch((err) => {
    console.error(err);
    process.exit(1);
});
